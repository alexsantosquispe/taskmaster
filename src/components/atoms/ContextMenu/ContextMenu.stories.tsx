import { EllipsisVerticalIcon } from '@/icons';
import { ContextMenu } from './ContextMenu';

import { Wrapper } from '@/utils/wrappers.utils';
import type { Meta, StoryObj } from '@storybook/react';
const options = [
  { label: 'Option 1', value: 'option-1' },
  { label: 'Option 2', value: 'option-2' },
  { label: 'Option 3', value: 'option-3' },
  { label: 'Option 4', value: 'option-4' }
];

const meta = {
  title: 'Atoms/ContextMenu',
  component: ContextMenu,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof ContextMenu>;

export default meta;

type Story = StoryObj<typeof ContextMenu>;

export const Default: Story = () => {
  return (
    <div className="grid grid-cols-2 gap-8">
      <Wrapper>
        <ContextMenu
          ariaLabel="test"
          options={options}
          onSelectOption={() => {}}
          label="Menu options"
          className={{
            contentWrapper: 'min-w-[7rem]',
            button: 'border border-neutral-200 px-2'
          }}
        />
      </Wrapper>

      <Wrapper>
        <ContextMenu
          ariaLabel="test"
          options={options}
          onSelectOption={() => {}}
          icon={<EllipsisVerticalIcon />}
          className={{
            contentWrapper: 'min-w-[7rem]',
            button: 'border border-neutral-200 px-2'
          }}
        />
      </Wrapper>
    </div>
  );
};

Default.args = {};
