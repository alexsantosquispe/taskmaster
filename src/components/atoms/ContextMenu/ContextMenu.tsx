import { useClickOutside } from '@/hooks/useClickOutside';
import type { Option } from '@/models/types';
import { useRef, useState, type ReactNode } from 'react';
import { twMerge } from 'tailwind-merge';
import { Button } from '../Button/Button';

interface ContextMenuProps {
  ariaLabel: string;
  options: Option[];
  onSelectOption: (option: Option) => void;
  label?: string;
  icon?: ReactNode;
  isSecondary?: boolean;
  className?: {
    mainContainer?: string;
    button?: string;
    contentWrapper?: string;
  };
}

export const ContextMenu = ({
  ariaLabel,
  options,
  onSelectOption,
  label = '',
  icon,
  isSecondary = true,
  className
}: ContextMenuProps) => {
  const [isOpen, setIsOpen] = useState(false);
  const menuRef = useRef<HTMLDivElement>(null);

  useClickOutside(menuRef, () => setIsOpen(false));

  const toggleMenu = () => setIsOpen((prev) => !prev);

  const onSelectOptionHandler = (option: Option) => {
    onSelectOption(option);
    setIsOpen(false);
  };

  return (
    <div
      ref={menuRef}
      className={twMerge('relative flex flex-col', className?.mainContainer)}
    >
      <Button
        ariaLabel={ariaLabel}
        onClick={toggleMenu}
        isSecondary={isSecondary}
        label={label}
        icon={icon}
        className={twMerge('rounded-md p-1', className?.button)}
      />
      {isOpen && (
        <div
          data-testid="context-menu"
          className={twMerge(
            'dark:bg-primary absolute top-full right-0 mt-1 w-fit rounded-lg border border-neutral-200 bg-white p-1 shadow-md dark:border-white/20 dark:shadow-none',
            className?.contentWrapper
          )}
        >
          <ul className="flex flex-col gap-1">
            {options.map((option) => (
              <li key={option.value}>
                <Button
                  label={option.label}
                  ariaLabel={`${option.label} option`}
                  onClick={() => onSelectOptionHandler(option)}
                  isSecondary={true}
                  className="hover:text-primary w-full justify-start rounded-md px-2 py-1 font-medium hover:cursor-default dark:hover:text-white"
                />
              </li>
            ))}
          </ul>
        </div>
      )}
    </div>
  );
};
