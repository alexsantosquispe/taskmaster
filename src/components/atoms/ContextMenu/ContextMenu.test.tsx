import '@testing-library/jest-dom';

import { fireEvent, render, screen } from '@testing-library/react';

import type { Option } from '@/models/types';
import { ContextMenu } from './ContextMenu';

const options: Option[] = [
  { label: 'Edit', value: 'edit' },
  { label: 'Delete', value: 'delete' },
  { label: 'Archive', value: 'archive' },
  { label: 'Unarchive', value: 'unarchive' }
];

describe('ContextMenu', () => {
  const onSelectOptionMock = jest.fn();
  const props = {
    ariaLabel: 'test',
    options,
    onSelectOption: onSelectOptionMock,
    label: 'Menu'
  };

  describe('styles', () => {
    it('should render the context menu correctly', () => {
      const component = render(<ContextMenu {...props} />);

      expect(component).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    beforeEach(() => {
      render(
        <div data-testid="main-wrapper">
          <ContextMenu {...props} />
        </div>
      );
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('should display the menu options when click the menu button', () => {
      const menuButton = screen.getByText('Menu');

      expect(menuButton).toBeInTheDocument();

      fireEvent.click(menuButton);

      expect(screen.getByTestId('context-menu')).toBeInTheDocument();

      const menuItem = screen.getAllByRole('listitem');

      expect(menuItem).toHaveLength(options.length);
    });

    it('should hide the menu options when select and option', () => {
      fireEvent.click(screen.getByText('Menu'));

      expect(screen.getByTestId('context-menu')).toBeInTheDocument();

      const option = screen.getByText('Edit');

      fireEvent.click(option);

      expect(onSelectOptionMock).toHaveBeenCalledTimes(1);

      expect(screen.queryByTestId('context-menu')).not.toBeInTheDocument();
    });

    it('should hide the menu options when click the menu button again', () => {
      fireEvent.click(screen.getByText('Menu'));

      expect(screen.getByTestId('context-menu')).toBeInTheDocument();

      fireEvent.click(screen.getByText('Menu'));

      expect(screen.queryByTestId('context-menu')).not.toBeInTheDocument();
    });

    it('should hide when click outside the menu', () => {
      fireEvent.click(screen.getByText('Menu'));

      expect(screen.getByTestId('context-menu')).toBeInTheDocument();

      fireEvent.mouseDown(document.body);

      expect(screen.queryByTestId('context-menu')).not.toBeInTheDocument();
    });
  });
});
