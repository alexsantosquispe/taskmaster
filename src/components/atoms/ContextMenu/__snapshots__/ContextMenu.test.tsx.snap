// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ContextMenu styles should render the context menu correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative flex flex-col"
      >
        <button
          aria-label="test"
          class="flex w-fit items-center justify-center gap-2 text-sm font-semibold transition-colors duration-200 ease-in-out text-primary dark:text-white hover:text-accent dark:hover:text-accent-dark hover:cursor-pointer hover:bg-neutral-100 dark:hover:bg-white/15 rounded-md p-1"
          type="button"
        >
          Menu
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative flex flex-col"
    >
      <button
        aria-label="test"
        class="flex w-fit items-center justify-center gap-2 text-sm font-semibold transition-colors duration-200 ease-in-out text-primary dark:text-white hover:text-accent dark:hover:text-accent-dark hover:cursor-pointer hover:bg-neutral-100 dark:hover:bg-white/15 rounded-md p-1"
        type="button"
      >
        Menu
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
